### DISCUSSION TITLE
Python in visidata

### DISCUSSION THREAD

**Discussion by Shahin-rmz** [saulpw/visidata#1294] [Q&A] [answered]
Hello,
the more I use Visidata, the more fall in love with it.
2 questions, maybe newbe ones.
can not find good documentation where shows how to use Python inside Visidata. Could someone please show me any links if there are any?
2. how can I choose not null values inside a column?
thanks for your time

**Comment by anjakefala** [comment, marked as answer]
These are not newbe questions at all!

For the second one, the answer can be found here: https://www.visidata.org/docs/rows/

> How to select rows where the current column is not null or empty?

> 'Null' cells, by default, are cells which contain `None`. This can be changed with `options.null_value`. Null cells can be set with `zd` (set current cell to `None`) or `gzd` (set selected rows in current column to `None`). Null cells are distinguished with a yellow ∅' symbol on the right hand corner. They are distinct from empty cells (which are `''` in columns of string type.)
 > Type `|` followed by `.` to select all rows without empty or null cells in the current column.

**Reply by Shahin-rmz** [reply]
Thanks @anjakefala for your answer.
with your help and the documentation, I could manage to change the `#Err` to  `∅` symbol. how ever, wanted to know if there is better way to do it than manually select them via `|` and regex and then do `gzd`.
But your answer was intuitive.
Thanks

**Comment by anjakefala** [comment]
For the first, I do not think there is good comprehensive documentation about this, yet. 

Essentially, there are commands that accept *regex expressions* and *Python expressions*. For the ones that accept *Python expressions*, you can pass column names as variables, and the expression will be performed on every row in that column.

For example, `=` (`expr-col`), you could type (`Column_A + 2`). This will create a new column. For each row in `Column_A`, the contents will be `+2`, and then the solution will be placed in that same row in the new column. A good overview of this command can be found here: https://jsvine.github.io/intro-to-visidata/intermediate/creating-new-columns/#how-to-create-an-expression-column.

Other example commands that accept Python expressions are `z|` (select by Python expression), `z/` (search by Python expression),  `Ctrl+X` and friends (evaluate the Python expression, and open a new PythonObj sheet). It would be good to have documentation for these.

To learn which commands accept a Python expression as input,  you can open the manpage (`g Ctrl+H` within VisiData), then type `/ python`, and scroll through with `n`/`N`. =)

**Reply by Shahin-rmz** [reply]
Thanks @anjakefala for your answers. actually what I meant was opening vim within visidata, and then, define a function like that,
```python
def awesome_visidata(current_col):
    if foo :
        do some advanced functions
    elif bar:
        do some other advanced functions
```
By Python within visidata, I meant that. not python regex expressions :)

**Reply by saulpw** [reply]
It's a good idea, @Shahin-rmz!  This is not quite possible, though you can use `exec-python` and then `Ctrl+O` to launch your $EDITOR.  It just doesn't get saved into the globals dict like you'd expect :)  Maybe someday.

**Reply by saulpw** [reply]
Okay, now symbols defined in `exec-python` are saved into the globals dict, and will be available for use in expressions!  So you can do:

1. `<space>exec-python<enter>` to start the command (there is no keybinding, but you can also find it under System > Python > exec())
2. `Ctrl+O` to open the input in your `$EDITOR` (if set).
3. Type out the code you want, save and quit when done.
4. Press `<enter>` at the input prompt to accept the input.
5. Now you should be able to use the definitions in subsequent expressions!

**Reply by Shahin-rmz** [reply]
It's simply fabulous.
with your help, I could manage to open my editor (which is of course Vim).

two things to mention though, 
- the editor behaves the code as plain text, not python script. should I manually add `#!/usr/bin/env python3` to the beginning?

- I want to know how you perform functions on the columns (it is not pandas and I don't know how to call columns and do the functions), if you could link an example it would be awesome.


thanks

**Reply by saulpw** [reply]
To use a function that you define, use `=awesome_visidata(current_col)`.

For the other question, vim has a filetype setting you can pass as an argument.

**Comment by ghost** [comment]
@Shahin-rmz Have a look at these discussions where I've been trying to understand similar things. I don't know if you can define a function 'live' inside visidata, but you can certainly define a function in your .visidatarc which will load when you start visidata and be available for you to use inside Python expressions. If you get more serious, you can also create Python plugins:

- [Adding Python functions to ~/.visidatarc](https://github.com/saulpw/visidata/discussions/1117)
- [Puzzling out plugin creation](https://github.com/saulpw/visidata/discussions/984)
